File: Data Structures & Algorithms/string-encode-and-decode/submission-0.java
class Solution {

    public String encode(List<String> strs) {
        StringBuilder sb = new StringBuilder();
        for (String str : strs) {
            sb.append((char)str.length()).append(str);
        }
        return sb.toString();
    }

    public List<String> decode(String str) {
        List<String> result = new ArrayList<>();
        int i = 0, n = str.length();
        while (i < n) {
            int size = (int) str.charAt(i++);
            result.add(str.substring(i, i + size));
            i += size;
        }
        return result;
    }
}

File: Data Structures & Algorithms/count-subsequences/submission-0.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        if len(t) > len(s):
            return 0

        dp = {}

        def dfs(i, j):
            if j >= len(t):
                return 1
            if i >= len(s):
                return 0

            if (i, j) in dp:
                return dp[(i, j)]

            result = 0
            if s[i] == t[j]:
                result = dfs(i + 1, j) + dfs(i + 1, j + 1)
            else:
                result = dfs(i + 1, j)

            dp[(i, j)] = result
            return result

        return dfs(0, 0)

        